linux/src/threadpool_linux.c

```diff
@@ -217,29 +195,23 @@ static int threadpool_work_func(void* param)
                     if (work_function != NULL)
                     {
                         work_function(work_function_ctx);
-                        /* Codes_SRS_THREADPOOL_LINUX_05_039: [ threadpool_work_func shall acquire the shared SRW lock by calling srw_lock_acquire_shared. ]*/
+
                         srw_lock_acquire_shared(threadpool->srw_lock);
-                        /* Codes_SRS_THREADPOOL_LINUX_05_040: [ threadpool_work_func shall save pending_work_item_count_ptr is not NULL in is_pending_work_item_count_ptr_not_null. ]*/
                         bool is_pending_work_item_count_ptr_not_null = (NULL != threadpool->task_array[current_index].pending_work_item_count_ptr);
-                        /* Codes_SRS_THREADPOOL_LINUX_05_041: [ threadpool_work_func shall release the shared SRW lock by calling srw_lock_release_shared. ]*/
                         srw_lock_release_shared(threadpool->srw_lock);
-                        /* Codes_SRS_THREADPOOL_LINUX_05_042: [ If the is_pending_work_item_count_ptr_not_null is TRUE then: ]*/
+
                         if (is_pending_work_item_count_ptr_not_null)
-                        {
-                            /* Codes_SRS_THREADPOOL_LINUX_05_043: [ threadpool_work_func shall acquire the exclusive SRW lock by calling srw_lock_acquire_exclusive. ]*/
+                        {                            
                             srw_lock_acquire_exclusive(threadpool->srw_lock);
-                            /* Codes_SRS_THREADPOOL_LINUX_05_044: [ threadpool_work_func shall decrement the pending_work_item_count_ptr by calling interlocked_decrement. ]*/
-                            interlocked_decrement(threadpool->task_array[current_index].pending_work_item_count_ptr);
-                            /* Codes_SRS_THREADPOOL_LINUX_05_045: [ threadpool_work_func shall send wake up signal to single listener for the address in pending_work_item_count_ptr. ]*/
-                            wake_by_address_single(threadpool->task_array[current_index].pending_work_item_count_ptr);
-                            /* Codes_SRS_THREADPOOL_LINUX_05_046: [ threadpool_work_func shall release the shared SRW lock by calling srw_lock_release_exclusive. ]*/
+                            (void)interlocked_decrement(threadpool->task_array[current_index].pending_work_item_count_ptr);
+                            wake_by_address_single(threadpool->task_array[current_index].pending_work_item_count_ptr);                         
```

Comment:
Since you're only waiting up if the value is zero, then you need to have an if on the decrement as to only wait on zero, it makes it more efficient.

if (interlocked_decrement(...) == 0)
{
     wake_by_address_single(...);
}

Reply:
Okay, generally if the code changes or expands, and we add more listeners to this address; like do something when the pending work item count becomes 100, wouldn't this constraint create an issue and cause major debug issue? I know in this case we are waiting on 0 but in general would it be a big over head to send these signals?

Reply:
Fixed